haystack/preview/document_stores/memory/document_store.py

```diff
@@ -142,3 +160,59 @@ def delete_documents(self, document_ids: List[str]) -> None:
             if not doc_id in self.storage.keys():
                 raise MissingDocumentError(f"ID '{doc_id}' not found, cannot delete it.")
             del self.storage[doc_id]
+
+    def bm25_retrieval(
+        self, query: str, filters: Optional[Dict[str, Any]] = None, top_k: int = 10, scale_score: bool = True
+    ) -> List[Document]:
+        """
+        Retrieves documents that are most relevant to the query using BM25 algorithm.
+
+        :param query: The query string.
+        :param filters: A dictionary with filters to narrow down the search space.
+        :param top_k: The number of top documents to retrieve. Default is 10.
+        :param scale_score: Whether to scale the scores of the retrieved documents. Default is True.
+        :return: A list of the top 'k' documents most relevant to the query.
+        """
+        if not query:
+            raise ValueError("Query should be a non-empty string")
+
+        filters = filters or {}
+        default_filters = {"content_type": ["text", "table"]}
+        final_filters = {**filters, **default_filters}
```

Comment:
Yes, that's it! Even better, I'd fail if the users tried to set a content type filter that is incompatible, something like (**untested**, it's probably missing an `Iterable` import too):

```python
if filters and "content_type" in filters:
    filtered_content_types = filters["content_type"]
    if isinstance(filtered_content_types, str) and filtered_content_types not in ["text", "table"]:
        raise ValueError("MemoryDocumentStore can perform BM25 retrieval on no other document type than text or table.")
    elif isinstance(filtered_content_types, Iterable) and set(filtered_content_types).issubset({"text", "table"}):
        raise ValueError("MemoryDocumentStore can perform BM25 retrieval on no other document type than text or table.")
else:
    filters = {**filters, "content_type": ["text", "table"]}
```

The only difference between this code and your is that in this way I let the user select only text or only table, if they want to.

Reply:
Let's also add a few tests for this, to check out the various combinations:

- `filters = {}`
- `filters = {'content_type': 'text'}`
- `filters = {'content_type': ['text']}`
- `filters = {'content_type': 'audio'}`
- `filters = {'content_type': ['audio']}`
- `filters ={'content_type': ['text', 'table']}`
- `filters ={'content_type': ['table', 'audio']}`